pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,175 @@
+		},
+		{
+			description: "Fails registration with conflicting plugin name during initial traverse after plugin watcher restarts",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+	}
+
+	callbackCount := 0
+	w.AddHandler("example-plugin-type", func(name string, versions []string, sockPath string) error {
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, "example-plugin", name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithBlock(),
+			grpc.WithTimeout(10*time.Second),
+			grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+				return net.DialTimeout("unix", addr, timeout)
+			}))
+		require.Nil(t, err)
+		defer conn.Close()
+
+		// The plugin handler should be able to use any listed service API version.
+		v1beta1Client := v1beta1.NewExampleClient(conn)
+		v1beta2Client := v1beta2.NewExampleClient(conn)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+		require.Nil(t, err)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+		require.Nil(t, err)
+		ret := testCases[callbackCount].returnErr
+		callbackCount++
+		return ret
+	})
+	err = w.Start()
+	require.NoError(t, err)
+	for {
+		if w.watcher != nil {
+			break
+		}
+	}
+
+	p := NewTestExamplePlugin() //first cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	status := <-p.registrationStatus
+	require.True(t, status.PluginRegistered)
+
+	// Trying to start a plugin service at the same socket path should fail
+	// with "bind: address already in use"
+	err = p.Serve(socketPath)
+	require.NotNil(t, err)
+
+	// grpcServer.Stop() will remove the socket and starting plugin service
+	// at the same path again should succeeds and trigger another callback.
+	p.Stop()
+	p = NewTestExamplePlugin() // 2nd cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	status = <-p.registrationStatus
+	require.True(t, status.PluginRegistered)
+
+	// Starting another plugin with the same name got verification error.
+	p2 := NewTestExamplePlugin() //3rd cbk
+	socketPath2 := socketDir + "/plugin2.sock"
+	go func() {
+		p2.Serve(socketPath2)
+	}()
+	status = <-p2.registrationStatus
+	require.False(t, status.PluginRegistered)
+
+	// Restarts plugin watcher should traverse the socket directory and issues a
+	// callback for every existing socket.
+	glog.Infof("stopping watcher 1\n")
+	err = w.Stop()
+	require.NoError(t, err)
+	errCh := make(chan error)
+	go func() {
+		glog.Infof("starting watcher 2\n")
+		errCh <- w.Start()
+		glog.Infof("started watcher 2\n")
+	}()
+	select {
+	case status = <-p.registrationStatus:
+	case <-time.After(time.Second):
```

Comment:
Similar question here. Is 1 sec enough? Mostly just personal preference, I usually prefer to just using "status = <-p.registrationStatus" in unit tests for simplicity, but if you feel using timeout bound wait is helpful, maybe we can define a helper func and use it consistently in this file?